algorithm.go

```diff
@@ -1,22 +1,37 @@
 package httpsignatures
 
 import (
-	"crypto/sha1"
-	"crypto/sha256"
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/rsa"
 	"errors"
-	"hash"
+	"math/big"
 )
 
 var (
-	AlgorithmHmacSha256 = &Algorithm{"hmac-sha256", sha256.New}
-	AlgorithmHmacSha1   = &Algorithm{"hmac-sha1", sha1.New}
+	AlgorithmHmacSha256  = &Algorithm{"hmac-sha256", hmacSign(crypto.SHA256), hmacVerify(crypto.SHA256)}
+	AlgorithmHmacSha1    = &Algorithm{"hmac-sha1", hmacSign(crypto.SHA1), hmacVerify(crypto.SHA1)}
+	AlgorithmRsaSha256   = &Algorithm{"rsa-sha256", rsaSign(crypto.SHA256), rsaVerify(crypto.SHA256)}
+	AlgorithmRsaSha1     = &Algorithm{"rsa-sha1", rsaSign(crypto.SHA1), rsaVerify(crypto.SHA1)}
+	AlgorithmEcdsaSha256 = &Algorithm{"ecdsa-sha256", ecdsaSign(crypto.SHA256), ecdsaVerify(crypto.SHA256)}
```

Comment:
This should probably validate that the key is using a P256 curve. If someone generates a P224 curve, the hash would be truncated when siging:

> Sign signs a hash (which should be the result of hashing a larger message) using the private key, priv. If the hash is longer than the bit-length of the private key's curve order, the hash will be truncated to that length. It returns the signature as a pair of integers. The security of the private key depends on the entropy of rand.

https://golang.org/pkg/crypto/ecdsa/#PrivateKey

Reply:
Technically "ecdsa-sha256" only prescribes that the digest function uses SHA-256. The EC key can use any curve or number of bits. That's something the consumer must be able to infer from the key id.